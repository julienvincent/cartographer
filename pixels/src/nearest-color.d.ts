declare module 'nearest-color' {
  export type RGBColor = {
    r: number;
    g: number;
    b: number;
  };
  export type InputColors = Record<string, RGBColor>;

  export type ColorFinder = (
    color: RGBColor | string
  ) => {
    name: string;
    color: string;
    rgb: RGBColor;
  };
  export function from(colors: InputColors): ColorFinder;
}
