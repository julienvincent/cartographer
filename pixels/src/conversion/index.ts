export * from './pixel-data';
export * from './colors';
