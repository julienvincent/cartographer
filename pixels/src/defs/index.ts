export type RGBColor = [number, number, number];

export type Pixel = {
  r: number;
  g: number;
  b: number;
};

export type PixelGrid = Pixel[][];

export type BlockProperties = Record<string, string> & {
  falling?: boolean;
  liquid?: boolean;
};
export type MCBlockDefinition = {
  id: string;
  properties?: BlockProperties;
};

export type MCBlockWithOffset = MCBlockDefinition & {
  y_offset: number;
};

export type MCBlockGrid = MCBlockWithOffset[][];

export type BlockPaletteItem = MCBlockDefinition & {
  colors: RGBColor[];
};
export type BlockPalette = BlockPaletteItem[];

export type ColorPaletteItem = {
  id: string;
  colors: RGBColor[];
  blocks: MCBlockDefinition[];
};
export type ColorPalette = ColorPaletteItem[];
