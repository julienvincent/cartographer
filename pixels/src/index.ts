export * as conversion from './conversion';
export * as defs from './defs';
