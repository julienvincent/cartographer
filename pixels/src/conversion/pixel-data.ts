import * as defs from '../defs';
import * as _ from 'lodash';

function average<T>(datums: T[], getter: (datum: T) => number) {
  const sum = datums.reduce((acc, datum) => {
    return acc + getter(datum);
  }, 0);
  return Math.ceil(sum / datums.length);
}

/**
 * Convert a given browser-native ImageData object into a grid of `Pixel`s
 *
 * This essentially converts a flat array of numbers sequentially ordered as r,g,b,a into a
 * multidimensional array of Pixel maps.
 *
 * [r,g,b,a,  r,g,b,a,  r,g,b,a,  r,g,b,a]
 * ->>
 * [[Pixel, Pixel], [Pixel, Pixel]]
 *
 * This makes working with the data in later steps much easier
 */
export const convertImageDataToPixelGrid = (
  image_data: ImageData
): defs.PixelGrid => {
  const pixels = _.chunk(image_data.data, 4).map((pixel) => {
    const [r, g, b] = pixel;
    return {
      r,
      g,
      b,
    };
  });

  return _.chunk(pixels, image_data.width);
};

export const convertPixelGridToImageData = (pixel_grid: defs.PixelGrid) => {
  const width = pixel_grid[0].length;
  const height = pixel_grid.length;

  return pixel_grid.reduce((image_data, pixels, h) => {
    return pixels.reduce((image_data, pixel, w) => {
      const i = h * height * 4 + w * 4;
      const { r, g, b } = pixel;
      image_data.data[i] = r;
      image_data.data[i + 1] = g;
      image_data.data[i + 2] = b;
      image_data.data[i + 3] = 255;
      return image_data;
    }, image_data);
  }, new ImageData(width, height));
};

/**
 * Attempt to scale a given pixel grid down to a given target width/height. This works by creating
 * quadrants in the source pixel grid and averaging the pixels in each quadrant together
 */
export const scaleDownPixelGrid = (
  pixel_grid: defs.PixelGrid,
  width: number,
  height: number
) => {
  const source_width = pixel_grid[0].length;
  const source_height = pixel_grid.length;
  const width_pixel_scale = Math.ceil(source_width / width);
  const height_pixel_scale = Math.ceil(source_height / height);

  return _.range(height).map((target_height) => {
    return _.range(width).map((target_width) => {
      const pixels: defs.Pixel[] = [];

      for (let diff_h = 0; diff_h < height_pixel_scale; diff_h++) {
        for (let diff_w = 0; diff_w < width_pixel_scale; diff_w++) {
          const pixel_h = target_height * height_pixel_scale + diff_h;
          const pixel_w = target_width * width_pixel_scale + diff_w;
          const pixel = pixel_grid[pixel_h][pixel_w];
          pixels.push(pixel);
        }
      }

      return {
        r: average(pixels, (pixel) => pixel.r),
        g: average(pixels, (pixel) => pixel.g),
        b: average(pixels, (pixel) => pixel.b),
      };
    });
  });
};

export const scaleUpPixelGrid = (
  pixel_grid: defs.PixelGrid,
  width: number,
  height: number
) => {
  const source_width = pixel_grid[0].length;
  const source_height = pixel_grid.length;
  const width_pixel_scale = width / source_width;
  const height_pixel_scale = height / source_height;

  return _.range(height).map((target_height) => {
    return _.range(width).map((target_width) => {
      const pixel_h = Math.floor(target_height / height_pixel_scale);
      const pixel_w = Math.floor(target_width / width_pixel_scale);
      return pixel_grid[pixel_h][pixel_w];
    });
  });
};
