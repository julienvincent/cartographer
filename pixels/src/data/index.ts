import color_palette from './color-palette-1.19.json';
import * as defs from '../defs';

export const MC_BLOCK_COLORS = color_palette as unknown as defs.ColorPalette;
