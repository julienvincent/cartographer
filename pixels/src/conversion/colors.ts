import { MC_COLOR_MAP } from '../data/mc-color-mapping';
import * as nearest from 'nearest-color';
import * as defs from '../defs';

const findClosestMatchingColor = nearest.from(
  MC_COLOR_MAP.reduce((acc: Record<string, defs.Pixel>, { colors }, i) => {
    const [[r, g, b]] = colors;
    acc[i] = {
      r,
      g,
      b
    };

    return acc;
  }, {})
);

export const convertPixelGridColorsForMC = (pixel_grid: defs.PixelGrid) => {
  return pixel_grid.map((pixels) => {
    return pixels.map((pixel) => {
      const { rgb } = findClosestMatchingColor(pixel);
      return rgb;
    });
  });
};

export const applyGrayScale = (pixel_grid: defs.PixelGrid) => {
  return pixel_grid.map((pixels) => {
    return pixels.map((pixel) => {
      const { r, g, b } = pixel;
      const average = (r + g + b) / 3;
      return {
        r: average,
        g: average,
        b: average
      };
    });
  });
};
