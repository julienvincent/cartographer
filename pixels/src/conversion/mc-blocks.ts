import { MC_BLOCK_COLORS } from '../data';
import * as defs from '../defs';

const findColorIndex = (colors: defs.RGBColor[], pixel: defs.Pixel) => {
  return colors.findIndex((color) => {
    const [r, g, b] = color;
    return r === pixel.r && g === pixel.g && b === pixel.b;
  });
};

export const convertPixelGridToMCBlocks = (pixel_grid: defs.PixelGrid): defs.MCBlockGrid => {
  return pixel_grid.map((pixels) => {
    return pixels.map((pixel) => {
      const matching = MC_BLOCK_COLORS.find((mapping) => {
        return findColorIndex(mapping.colors, pixel) !== -1;
      });

      const [block_id] = matching!.block_ids;

      return {
        block_id: block_id,
        height_variant: findColorIndex(matching!.colors, pixel)
      };
    });
  });
};
