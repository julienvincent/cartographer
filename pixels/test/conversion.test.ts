import * as conversion from '../src/conversion';
import 'jest';
import {test, expect, describe} from '@jest/globals'

describe('pixel conversions', () => {
  test('it should correctly convert ImageData to a pixel grid', () => {});

  test('it should correctly scale a PixelGrid down', () => {
    console.log('done!');

    const arr = [1,2]

    expect(arr).toMatchSnapshot()
  });
});
