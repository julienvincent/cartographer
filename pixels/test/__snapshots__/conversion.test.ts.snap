// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`pixel conversions it should correctly scale a PixelGrid down 1`] = `
Array [
  1,
  2,
]
`;
